Feature: 강의 목록 기능입니다.

park3hho opened the pull request:
@ ### ✅ PR 요약
- 관련 이슈 번호: #123
- 작업 요약: 레디스를 활용한 목록 조회기능 

## 📄 상세 내용
- [1] 테스트 코드 작성
- [2] 시리얼라이저 수정
## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [1] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [2] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).
#

park3hho committed:
✅test: <test 파일 생성입니다.>

park3hho committed:
✅test: <test 파일 작성입니다.>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] ✅ test_lecture_title_is_correct 함수 생성
[2] ✅ setUp 함수 생성
[3] ✅ test_lecture_creation 함수 생성

park3hho committed:
# 💡 build: <REQ-LECT-001, 강의 목록 조회 API>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 REQ-LECT-001, 강의 목록 조회 API 작성
[2] 💡
[3] 💡
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 0.1.1 #
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

# Conflicts:
#	apps/lectures/migrations/0004_alter_category_created_at_alter_category_updated_at_and_more.py

# Conflicts:
#	config/urls.py

# Conflicts:
#	config/urls.py

# Conflicts:
#	config/urls.py

park3hho committed:
# 💡 build: <REQ-LECT-001, 강의 목록 조회 API>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 REQ-LECT-001, 강의 목록 조회 API 작성
[2] 💡 REQ-LECT-001, 분류 기능 추가
[3] 💡 REQ-LECT-001, SwaggerUI에 명시
[4] 💡 REQ-LECT-001, 페이지네이션 추가

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 0.1.2 #
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# ✅ test: <REQ-LECT-001, 테스트 코드 추가 작성>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] ✅ REQ-LECT-001, 최신 순 정렬 기능에 따른 결과 테스트 추가
[2] ✅ REQ-LECT-001, 별점 순 정렬 기능에 따른 결과 테스트 추가
[3] ✅ REQ-LECT-001, 가격 순 정렬 기능에 따른 결과 테스트 추가
[4] ✅ REQ-LECT-001, 검색 기능에 따른 테스트 추가

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모
# cached_lecture_list.py 올리면 안됨
# 0.2.1 #
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# ✅ test: <REQ-LECT-001, 테스트 코드 추가 작성>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] ✅ REQ-LECT-001, 카테고리를 통해 검색 결과 표시
[2] ✅ REQ-LECT-001, 검색 결과에 따른 카테고리 표시
[3] ✅ REQ-LECT-001, 카테고리 없을 때 상황 테스트
[4] ✅ REQ-LECT-001, 검색 기능에 따른 테스트 추가

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.2 #
# cached_lecture_list.py 올리면 안됨

========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# 💡 chore: <black . & isort .>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 black . & isort .
[2] 💡
[3] 💡
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.3 #
# cached_lecture_list.py 올리면 안됨

========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# ✅ test: <When 및 Then 명시>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] ✅ test_lecture_to_categories
[2] ✅ test_category_to_lectures
[3] ✅ test_lecture2_no_wrong_categories
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.4 #
# cached_lecture_list.py 올리면 안됨

========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# ✨test: <Redis 사용하여 캐시된 데이터로 목록 조회 기능>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] ✨ Redis 사용하여 캐시된 데이터로 목록 조회
[2] ✨ Redis에 넣는 Scripts 작성 -> 나중에 자동화 (올리지 않음)
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.5 #
# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
- (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
- (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# 💡test: <black . & isort . & mypy .>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 black . & isort . & mypy .
[2] ✨
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.6 #
# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
- (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
- (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# 💡test: <url 통일>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 url 통일
[2] ✨
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.7 #
# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
- (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
- (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# 💡test: <redis 도입한 테스트코드 재작성>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 테스트코드 작성. TDD 위배해버렸다.
[2] ✨
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.8 #
# 다음부턴 테스트 코드 먼저 작성하자
# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
- (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
- (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

# Conflicts:
#	poetry.lock
#	pyproject.toml

park3hho committed:
# 💡test: <CICD>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 CICD
[2] ✨
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.8 #
# 다음부턴 테스트 코드 먼저 작성하자
#
# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
- (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
- (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# 💡test: <CICD>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 CICD
[2] ✨
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.2.8 #
# 다음부턴 테스트 코드 먼저 작성하자
#
# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
- (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
- (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# 💡✅test: <CICD>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡✅ CI, TEST 통과하게끔 재작성함.
[2] ✨
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.3.1 #
# 다음부턴 테스트 코드 먼저 작성하자
#
# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
- (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
- (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
# 💡✅test: <CICD>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡✅ CI, TEST 통과하게끔 재작성함.
[2] ✨
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
# 관련 이슈: #123

# 4. 메모 # 0.3.2 #
# 다음부턴 테스트 코드 먼저 작성하자
#
# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
- (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
- (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

park3hho committed:
💡chore: <상관 없는 코드 삭제>

# 2. 변경 또는 추가사항을 아래에 간략하게 작성하세요 ( 필수 )
[1] 💡 상관 없는 코드 삭제
[2] ✨
[3] ✨
[4] 💡

# 3. 이슈 ( 선택 )
관련 이슈: #123

# 4. 메모 # 0.3.2 #
# 다음부턴 테스트 코드 먼저 작성하자

# hahamemo.txt 올리면 안됨
# cached_lecture_list.py 올리면 안됨 (Scripts)
# paginate, sort_lectures, filter_lectures 어디에 분리? -> PR로 올려서 검토 받기
# Redis에 넣는 로직 자동화:
# - (1) Log 관련 기능 완성하면 사용자 관련 추천 서비스와 연동하여, 검색할 때마다 추천 강의를 캐시에 넣도록 수정해야함.
# - (2) 데이터베이스에 새로운 강의 목록이 추가되면 캐시 데이터를 수정해야함.
========================================================================
# Commit Convention
# ✨ feat     : 새로운 기능 추가
# 🐛 fix      : 버그 수정
# 💡 chore    : 기능 추가 없이 코드 수정 (오타, 주석 등)
# 📝 docs     : 문서 수정 (README 등)
# 🚚 build    : 빌드 관련 파일 수정
# ✅ test     : 테스트 코드 추가/변경 (프로덕션 코드 변경 없음)
# ♻️ refactor : 리팩터링 (기능 변화 없음)
# 🚑 hotfix   : 긴급 수정

SimJunSik commented on apps/lectures/views/lecture_list.py:
```diff
@@ -0,0 +1,122 @@
+            redis_client.set("lectures:list", json.dumps(lectures))
+
+        # Filtering & Sorting
+        lectures = self.filter_lectures(lectures, request)  # type: ignore
+        lectures = self.sort_lectures(lectures, request.query_params.get("ordering"))  # type: ignore
```
DB 에서 할 수 있는 것들은 전부 DB 에서 하는게 좋아요
여기 filter, sort 두가지 전부 DB 에서 해야 성능 이슈가 아주아주 덜 합니다

밑에 filter, sort 함수들에서 type ignore 를 해두셨는데, lectures 는 list[Lecture] 가 아니라 QuerySet 이 되어야 합니다

park3hho replied:
성능 비교해서 테스트해봤는데 
캐시의 직렬화된 데이터를 리스트된 상태에서 하는게 0.008초 좌우 나오고,
DB에서 하는건 0.011초 정도 나옵니다.

나중에 데이터 많아지면 다시 비교해서 적용해보겠습니다.

SimJunSik replied:
네 작은 데이터량이지만 일단 비교를 해보셨다니 좋은데 단순 실행 시간만 보실게 아니라 더 생각해보시면 좋을 것 같습니다
성능이라는게 실행 시간만 있는게 아니라서요

비지니스 로직상 어쩔 수 없이 Application 단에서 하는게 아니면 실무에서는 DB 에서 하라고 피드백을 무조건 할거고,
면접 때 여기서 `왜 Application 단에서 필터, 정렬을 하셨나요?` 라는 질문에 지금처럼 답변을 하신다면 이런저런 꼬리 질문이 많이 들어올겁니다

SimJunSik commented on apps/lectures/views/lecture_list.py:
```diff
@@ -0,0 +1,122 @@
+        try:
+            redis_client = redis.Redis(host="redis", port=6379, db=0, decode_responses=True)
+            cached_lectures = redis_client.get("lectures:list")
+        except redis.exceptions.ConnectionError:
+            return Response({"detail": "Redis connection error"}, status=500)
```
memory DB, cache 에 문제가 있다고 해서 바로 에러를 반환하기 보다는 거기에 문제가 있다면 fallback 으로 DB 에서 가져와서 응답해주는게 더 좋을 것 같습니다

park3hho replied:
넵 확인했습니다.

Meoyoug commented on apps/lectures/serializers/crawled_lecture.py:
```diff
@@ -0,0 +1,31 @@
+
+
+class LectureSerializer(ModelSerializer[Lecture]):
+    categories = serializers.SerializerMethodField()
```
SerializerMethodField는 swagger-ui에 반영되지 않아서 별도로 extend_schema에 정의해줘야하는 번거로움이 있습니다
상기에 CategorySerializer 가 있으니 NestedSerializer로 구현해보는게 어떨 까요?

park3hho replied:
확인하고 수정했습니다. (삭제했어요. 원래 코드랑 상관이 없었어서,)

Meoyoug commented on apps/lectures/tests/test_lecture_list.py:
```diff
@@ -0,0 +1,219 @@
+    def test_get_lecture_list_with_cache(self, mock_redis: MagicMock) -> None:
+        # Given
+        fake_redis = mock_redis.return_value
+        cached_data = [{"title": "Cached Lecture", "instructor": "Charlie", "updated_at": "2025-09-01"}]
```
실제로 필요한 필드보다 키 - 밸류 쌍이 적네요 시리얼라이저가 주는 응답과 맞춰서 작성해주세요

park3hho replied:
넹

Meoyoug commented on apps/lectures/urls/urls.py:
```diff
@@ -0,0 +1,10 @@
+app_name = "lectures"
+
+urlpatterns = [
+    path("list/", LectureListView.as_view(), name="lecture_list"),
```
trailing slash 제거부탁해용

park3hho replied:
넵 수정했습니다.

Umdoong replied:
수정 안된 것 같아요..!!

Meoyoug commented on apps/lectures/tests/test_lecture_list.py:
테스트 작성 모킹을 활용하신건 너무 좋은 시도였습니다
다만 RedisTestClient를 만들어놓아서 Redis를 활용한 테스트가 필요하다면 해당 클래스 상속 받아서 진행해주세요`core/utils/test_clients.py` 에 있습니다

park3hho replied:
넵 확인했습니다.

Meoyoug commented on apps/lectures/views/lecture_list.py:
서비스 로직은 분리해서 services 모듈내에 구현해두기

Umdoong commented on apps/lectures/serializers/crawled_lecture.py:
```diff
@@ -0,0 +1,31 @@
+
+    class Meta:
+        model = Lecture
+        fields = (
```
duration, description field를 빼신 이유가 있나용?

park3hho replied:
요구사항 정의서에 따로 명시안되있는데 넣을까요

Umdoong replied:
추가했습니당 넣어주세요!

Umdoong commented on apps/lectures/serializers/crawled_lecture.py:
```diff
@@ -0,0 +1,31 @@
+        )
+
+    def get_categories(self, obj: Lecture) -> List[int]:
+        return [lc.category.id for lc in obj.lecturecategory_set.all()]
```
이거 N+1 문제 생기지 않나용?

park3hho replied:
지워떠영

Umdoong commented on apps/lectures/tests/test_lecture_list.py:
```diff
@@ -0,0 +1,219 @@
+
+
+class LectureTestCase(TestCase):
+    def setUp(self) -> None:
```
setUp의 경우 테스트 함수 실행 전에 계속 실행되기 때문에 다른 setup 방식도 찾아보시면 좋을 것 같아요
setUpTestData 같은 거요!

park3hho replied:
넹 확인했습니다.

park3hho committed:
✨ feat: 강의목록 기능 작성

- fallback 처리에 따른 테스트코드 작성
- fallback 처리
- 로깅 기능 구현
- 필요없는 코드 삭제

park3hho committed:
💡 chore: CICD 오류 수정

- CICD 오류 수정

# Conflicts:
#	poetry.lock

park3hho committed:
💡 chore: CICD 수정

park3hho committed:
🐛 fix: CICID

Umdoong commented on apps/recruitments/tests/tests_list.py:
```diff
@@ -98,7 +98,7 @@ def test_recruitment_list_get(self) -> None:
         res = self.client.get(url, data=query_params)
         self.assertEqual(res.status_code, 200)
-        print(res.data)
+        print(f"\n{res.data}")
```
프린트문 지워주세요

Umdoong commented on apps/lectures/views/lecture_list.py:
```diff
@@ -0,0 +1,132 @@
+        redis_client = None
+
+        try:
+            redis_client = redis.Redis(host="redis", port=6379, db=0, decode_responses=True)
```
settings/base.py에서 CACHES 설정을 레디스로 해놨기 때문에 단순 캐시만 활용할 거라면 그냥 from django.core.cache import cache 가져와서 사용해주시면 됩니다!

park3hho committed:
✨ feat     :  강의 목록 API

- Bookmark 매니저 수정
- 강의 serializer 수정
- 미사용 파일 삭제 (serializers/crawled_lectures.py) : 중복됨.
- 강의 목록 기능 리팩토링 (장고 내장 페이지네이션 사용, 매니저 사용, 캐시 설정 사용[원래는 직접 호출했었음], 서비스 로직 분리)
- 크롤링 테스트 추가 (100%)
- 매니저 로직에 맞게 테스트 코드 작성
- url 작성

park3hho committed:
💡 chore    : poetry.lock

# Conflicts:
#	poetry.lock

Umdoong commented on config/urls.py:
```diff
@@ -14,7 +14,8 @@
     path("api/v1/schedules", include("apps.study_group_schedules.urls")),
     path("api/v1/study-groups", include("apps.studies.urls")),
-    path("api/v1/lectures", include("apps.lectures.urls")),
+    path("api/v1/lectures", include("apps.lectures.urls.urls", namespace="lectures")),
```
이거 북마크 url origin/develop과 맞춰주세용

Umdoong commented on apps/lectures/managers/Lecture_manager.py:
```diff
@@ -0,0 +1,44 @@
+        names = [name.strip() for name in category_names.split(",")]
+        return self.filter(categories__name__in=names).distinct()
+
+    def sort_by_ordering(self, ordering: str) -> LectureQuerySet:
```
sort_by_ordering 이거 어디서 쓰고 있나용?

Umdoong commented on apps/lectures/views/lecture_list.py:
```diff
@@ -0,0 +1,62 @@
+    serializer_class = LectureSerializer
+    pagination_class = LectureCursorPagination
+
+    @method_decorator(cache_page(CACHE_TTL))
```
이거 URL path 기준으로 캐시돼서 Query Parameter가 바뀌어도 URL 경로가 같아서 캐시된 값을 그대로 반환할 거에용
https://docs.djangoproject.com/en/5.2/topics/cache/#the-per-view-cache
여기를 보시면 view의 응답을 캐시한다고 돼있어요

Meoyoug replied:
엇 효종님 이거 잘못된 내용입니다 코드 보시면 request.build_absolute_url()을 해서 캐쉬 키를 만들어요!
<img width="841" height="808" alt="image" src="https://github.com/user-attachments/assets/a20a4e83-716d-434c-af65-7c7a8ba17b56" />
<img width="694" height="450" alt="image" src="https://github.com/user-attachments/assets/d8fc5caa-7c02-4ca7-bad8-10943b3ffb27" />
<img width="669" height="185" alt="image" src="https://github.com/user-attachments/assets/eea39756-60c6-44d7-823e-3fa631b1e963" />

Umdoong replied:
오...공식문서만 보고 path param까지만 해주는 줄 알았는데, 코드를 뜯어보니 absolute_uri를 사용하는군요
그러면 query param까지 다 캐시가 되겠네요!!

park3hho committed:
💡 chore    : url 통합

url 안바꿨어서 했어여

park3hho committed:
💡 chore    : poetry 추가

park3hho committed:
🐛 fix      :  poetry.lcok Hash값 수정

park3hho committed:
💡 chore    : CICD 안되는 테스트 코드 문제

park3hho committed:
🐛 fix      : 테스트 코드 수정 중 (병렬 처리 오류)

park3hho committed:
💡 chore    : 병렬 문제 해결 중

park3hho committed:
💡 chore    : 병렬처리 오류 해결 중 3

park3hho committed:
💡 chore    : 병렬처리 문제 해결 중 4

park3hho committed:
💡 chore    : CICD 테스트 중

park3hho committed:
💡 chore    : poetry lock 수정
